pkg/apis/sources/v1beta1/apiserver_types.go

```diff
@@ -0,0 +1,149 @@
+*/
+
+package v1beta1
+
+import (
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/runtime"
+	"knative.dev/pkg/apis"
+	duckv1 "knative.dev/pkg/apis/duck/v1"
+	"knative.dev/pkg/kmeta"
+)
+
+// +genclient
+// +k8s:deepcopy-gen:interfaces=k8s.io/apimachinery/pkg/runtime.Object
+
+// ApiServerSource is the Schema for the apiserversources API
+type ApiServerSource struct {
+	metav1.TypeMeta   `json:",inline"`
+	metav1.ObjectMeta `json:"metadata,omitempty"`
+
+	Spec   ApiServerSourceSpec   `json:"spec,omitempty"`
+	Status ApiServerSourceStatus `json:"status,omitempty"`
+}
+
+// Check the interfaces that ApiServerSource should be implementing.
+var (
+	_ runtime.Object     = (*ApiServerSource)(nil)
+	_ kmeta.OwnerRefable = (*ApiServerSource)(nil)
+	_ apis.Validatable   = (*ApiServerSource)(nil)
+	_ apis.Defaultable   = (*ApiServerSource)(nil)
+	_ apis.HasSpec       = (*ApiServerSource)(nil)
+)
+
+const (
+	// ApiServerSourceAddEventType is the ApiServerSource CloudEvent type for adds.
+	ApiServerSourceAddEventType = "dev.knative.apiserver.resource.add"
+	// ApiServerSourceUpdateEventType is the ApiServerSource CloudEvent type for updates.
+	ApiServerSourceUpdateEventType = "dev.knative.apiserver.resource.update"
+	// ApiServerSourceDeleteEventType is the ApiServerSource CloudEvent type for deletions.
+	ApiServerSourceDeleteEventType = "dev.knative.apiserver.resource.delete"
+
+	// ApiServerSourceAddRefEventType is the ApiServerSource CloudEvent type for ref adds.
+	ApiServerSourceAddRefEventType = "dev.knative.apiserver.ref.add"
+	// ApiServerSourceUpdateRefEventType is the ApiServerSource CloudEvent type for ref updates.
+	ApiServerSourceUpdateRefEventType = "dev.knative.apiserver.ref.update"
+	// ApiServerSourceDeleteRefEventType is the ApiServerSource CloudEvent type for ref deletions.
+	ApiServerSourceDeleteRefEventType = "dev.knative.apiserver.ref.delete"
+)
+
+// +k8s:deepcopy-gen:interfaces=k8s.io/apimachinery/pkg/runtime.Object
+
+// ApiServerSourceList contains a list of ApiServerSource
+type ApiServerSourceList struct {
+	metav1.TypeMeta `json:",inline"`
+	metav1.ListMeta `json:"metadata,omitempty"`
+	Items           []ApiServerSource `json:"items"`
+}
+
+// ApiServerSourceSpec defines the desired state of ApiServerSource
+type ApiServerSourceSpec struct {
+	// inherits duck/v1 SourceSpec, which currently provides:
+	// * Sink - a reference to an object that will resolve to a domain name or
+	//   a URI directly to use as the sink.
+	// * CloudEventOverrides - defines overrides to control the output format
+	//   and modifications of the event sent to the sink.
+	duckv1.SourceSpec `json:",inline"`
+
+	// WatchResources is the list of resources to watch.
+	WatchResources []ApiServerResource `json:"watch"`
+
+	// ServiceAccountName is the name of the ServiceAccount to use to run the
+	// receive adapter for this source.
+	// +optional
+	ServiceAccountName string `json:"serviceAccountName,omitempty"`
```

Comment:
This field is exposing too much the underlying implementation. This assumes there is a receiver adapter per source. Today our sources only support sole-tenancy but maybe they should also be support multi-tenants (ie have only one receiver adapter installed in `knative-eventing`, the same way it's done for the dispatchers).

Reply:
Removed